transaction/beef.go

```diff
@@ -803,115 +800,214 @@ func (b *Beef) Verify(ctx context.Context, chainTracker chaintracker.ChainTracke
+		if len(bump.Path) > 0 {
+			// Check level 0 (leaf level) for transaction hashes
+			for _, elem := range bump.Path[0] {
+				if elem.Hash != nil && elem.Txid != nil && *elem.Txid {
+					txidsInBumps[elem.Hash.String()] = true
+				}
+			}
 		}
 	}
 
-	// Separate transactions that have at least one missing input
-	queue := make([]*BeefTx, 0)
-	for _, beefTx := range allTxs {
-		if beefTx.Transaction != nil {
-			hasMissing := false
-			for _, in := range beefTx.Transaction.Inputs {
-				if !validTxids[in.SourceTXID.String()] && b.findTxid(in.SourceTXID.String()) == nil {
-					missing[in.SourceTXID.String()] = true
-					hasMissing = true
-				}
+	result := &ValidationResult{
+		MissingInputs:     []string{},
+		NotValid:          []string{},
+		Valid:             []string{},
+		WithMissingInputs: []string{},
+		TxidOnly:          []string{},
+	}
+
+	// Maps for tracking
+	validTxids := make(map[string]bool)
+	txByID := make(map[string]*BeefTx)
+	missingInputs := make(map[string]bool)
+
+	// Lists for processing
+	var hasProof []*BeefTx
+	var txidOnly []*BeefTx
+	var needsValidation []*BeefTx
+	var withMissingInputs []*BeefTx
+
+	// First pass: categorize transactions
+	for txid, beefTx := range b.Transactions {
+		txByID[txid] = beefTx
+
+		switch beefTx.DataFormat {
+		case TxIDOnly:
+			// TxIDOnly transactions are valid if they appear in BUMPs
+			if txidsInBumps[txid] {
+				validTxids[txid] = true
+				txidOnly = append(txidOnly, beefTx)
+			} else {
+				// TxIDOnly without proof - add to txidOnly list but not valid
+				txidOnly = append(txidOnly, beefTx)
 			}
-			if hasMissing {
-				res.WithMissingInputs = append(res.WithMissingInputs, beefTx.Transaction.TxID().String())
+		case RawTxAndBumpIndex:
+			// Verify the bump index is accurate
+			if beefTx.BumpIndex >= 0 && beefTx.BumpIndex < len(b.BUMPs) {
+				bump := b.BUMPs[beefTx.BumpIndex]
+				// Check if this transaction appears in the specified bump
+				foundInBump := false
+				for _, elem := range bump.Path[0] {
+					if elem.Hash != nil && elem.Hash.String() == txid {
+						foundInBump = true
+						break
+					}
+				}
+				if foundInBump {
+					validTxids[txid] = true
+					hasProof = append(hasProof, beefTx)
+				} else {
+					// Invalid bump index - treat as needing validation
+					needsValidation = append(needsValidation, beefTx)
+				}
 			} else {
-				queue = append(queue, beefTx)
+				// Invalid bump index - treat as needing validation
+				needsValidation = append(needsValidation, beefTx)
+			}
+		case RawTx:
+			// RawTx is valid if it appears in a BUMP
+			if txidsInBumps[txid] {
+				validTxids[txid] = true
+				hasProof = append(hasProof, beefTx)
+			} else if beefTx.Transaction != nil {
+				// Check if all inputs are available
+				hasMissing := false
+				for _, input := range beefTx.Transaction.Inputs {
+					if _, exists := b.Transactions[input.SourceTXID.String()]; !exists {
+						missingInputs[input.SourceTXID.String()] = true
```

Comment:
Since calculating string txid is not zero-time, I suggest to precalculate it when it's used twice. 
```suggestion
				for _, input := range beefTx.Transaction.Inputs {
					sourceTxID := input.SourceTXID.String()
					if _, exists := b.Transactions[sourceTxID]; !exists {
						missingInputs[sourceTxID] = true
```